openml-lightgbm/lightgbm-provider/src/main/java/com/feedzai/openml/provider/lightgbm/LightGBMUtils.java

```diff
@@ -57,9 +57,11 @@ static public void loadLibs() {
 
         if (!libsLoaded) {
             try {
-                loadSharedLibraryFromJar("libgomp.so.1.0.0");
-                loadSharedLibraryFromJar("lib_lightgbm.so");
-                loadSharedLibraryFromJar("lib_lightgbm_swig.so");
+                final String cpuArchName = System.getProperty("os.arch");
+                final CpuArchitecture cpuArchitecture = CpuArchitecture.valueOf(cpuArchName.toUpperCase());
+                loadSharedLibraryFromJar("libgomp.so.1.0.0", cpuArchitecture);
+                loadSharedLibraryFromJar("lib_lightgbm.so", cpuArchitecture);
+                loadSharedLibraryFromJar("lib_lightgbm_swig.so", cpuArchitecture);
```

Comment:
Why are we including SO libraries within this jar? why aren't they separated artifacts?

Reply:
Well, if that is the approach... but IMO this is not correct. It will increase the jar size significantly and it is not taking advantage of the shared libraries of Operating Systems, this looks like statically linking these libraries. 

Another thing is that each time something is fixed on those libraries this jar will need to be created, etc...

Anyway, if this is the way that it is deployed, I leave that decision to the team.

Reply:
Again, it's how it was.  We are more concerned in making lgbm working on amd64 than in changing the structure how this is done. @AlbertoEAF can you give more context why this is done this way? Also @gandola the only thing I see that may be removed from here is the libgomp which we could use from the OS... right?

Reply:
Hi there :)

This is the best approach possible given our goals: low total cost of ownership, ease of deployment and reliability.  We put the dynamic libraries inside the (fat-)jar **on purpose** so it is self-sufficient in any deployment scenario.

This started for historical reasons and by design so we did not depending at all on the environment/setup of the target cluster/pulse machines used when we were mostly on-prem and clients managed those installations -- it would not be feasible at all to assume clients would properly manage the available shared libs on their cluster machines,  etc.; and if they did, the time for those changes to happen would be in the order of months, and the reliability of such setups questionable and prone to generating issues for our teams.

Now with the move to cloud this mechanism again proved simple and reliable. We didn't require coordination with other teams, and it always works. Additionally, with this strategy and from this MR onward we will be able to run on our cloud with both AMD64 and ARM64 machines without any configuration required, nor Pulse knowing at all whether the target EMR machines are AMD64 or ARM64.

We need to keep libgomp as well, as several of these machines don't have it available. This has been extensively tested on various clusters and without this library available at all times we'd be forced to fall back to a single thread for training (again, not viable).

The only requirement of this approach (so the .so's we ship are "universal") is that our compiled shared libraries be compatible with the C runtime of the target machines (glibc target machine >= glibc using during compile time of these .so's). That is why when we compile them we use an old docker environment that is older than any relevant linux distro in use.


Should we document this on the project @gandola ? Maybe add a summary of this to the docstring?